test/python/WMCore_t/WebTools_t/REST_t.py

```diff
@@ -141,11 +143,11 @@ def testException(self):
         list takes a single integer argument, querying with a string
         """
         url = self.urlbase + 'list?int=a'
-        self.assertRaises(urllib2.HTTPError, urllib2.urlopen, url)
+        self.assertRaises(urllib.error.HTTPError, urllib.request.urlopen, url)
         # urllib2,urlopen raise the error but not urllib.urlopen
         url = self.urlbase + 'list1?int=a'
         expected_data = {"exception": 400, "type": "HTTPError", "message": "Invalid input: Arguments added where none allowed"}
-        urllib_data = urllib.urlopen(url)
+        urllib_data = urllib.request.urlopen(url)
```

Comment:
Dear @amaltaro and @todor-ivanov,
This line causes the last unit test failure (actually this is the last "regression", we have other failures that have been present before this PR) that we have: https://cmssdt.cern.ch/dmwm-jenkins/view/All/job/DMWM-WMCore-PR-test/10248/testReport/WMCore_t.WebTools_t.REST_t/RESTTest/testException/

The cause seems to be the same as what we encountered with the previous unit test: `urllib.request.urlopen` raises an exception when there is a http error, instead of return data containing the details of the error.

This unit test is designed to check that the data that urlopen return in case of failures is what we expect.
This is now failing because this line (L150 - `urllib_data = urllib.request.urlopen(url)`) raises an exception before we check it.

if we change this unit test in thew following way:

```python
    def testException(self):
        """
        list takes a single integer argument, querying with a string
        """
        url = self.urlbase + 'list?int=a'
        self.assertRaises(urllib.error.HTTPError, urllib.request.urlopen, url)
        # urllib2,urlopen raise the error but not urllib.urlopen
        url = self.urlbase + 'list1?int=a'
-        expected_data = {"exception": 400, "type": "HTTPError", "message": "Invalid input: Arguments added where none allowed"}
+        # expected_data = {"exception": 400, "type": "HTTPError", "message": "Invalid input: Arguments added where none allowed"}
-        urllib_data = urllib.request.urlopen(url)
+        self.assertRaises(urllib.error.HTTPError, urllib.request.urlopen, url)
+        # urllib_data = urllib.request.urlopen(url)
-        if self.do_production:
-            #production mode returns 403 error
-            self.assertEqual(urllib_data.getcode(), 403)
-        else:
-            response_data = urllib_data.read()
-            response_data = json.loads(response_data)
-            self.assertEqual(response_data['type'], expected_data['type'])
-            self.assertEqual(response_data['message'], expected_data['message'])
-            self.assertEqual(urllib_data.getcode(), 400)
+        # if self.do_production:
+        #     #production mode returns 403 error
+        #     self.assertEqual(urllib_data.getcode(), 403)
+        # else:
+        #     response_data = urllib_data.read()
+        #     response_data = json.loads(response_data)
+        #     self.assertEqual(response_data['type'], expected_data['type'])
+        #     self.assertEqual(response_data['message'], expected_data['message'])
+        #     self.assertEqual(urllib_data.getcode(), 400)
```
Then the unit test succeeds because we catch the exception. This is not the best option because we only catch the exception, while the purpose of this unit test seems to be having a deeper introspection of the failure.

What can we do? What about something like the following?

```python
        # test 1: only check the exception that is raised
        self.assertRaises(urllib.error.HTTPError, urllib.request.urlopen, url)
        # test 2: checks the same exception as above, but introspects also the error code and message
        try:
            urllib.request.urlopen(url)
        except urllib.error.HTTPError as e:
            print("testException", str(e.hdrs), e.url)
            self.assertEqual(e.code, 400)
            self.assertEqual(e.msg, u'Bad Request')
            exception_data = json.loads(e.read())
            self.assertEqual(exception_data['type'], 'HTTPError')
            self.assertEqual(exception_data['message'], 'Invalid input: Arguments added where none allowed')
```

Reply:
Dario, this code snippet you provided looks good to me. It keeps the same test case. Thanks